### PR TITLE
INTMDB-263: Create Resource and Datasource for private_link_endpoint_adl

### PR DESCRIPTION
**Pull request by abner-dou**
## Description

A new resource and datasource was create for private link endpoints in datalake and online archive

Link to any related issue(s):[INTMDB-263](https://jira.mongodb.org/browse/INTMDB-263)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the Terraform contribution guidelines
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments
TESTS:

-----------------------------------------------------
--- PASS: TestAccResourceMongoDBAtlasPrivateLinkEndpointServiceADL_basic (6.90s)
PASS
coverage: 2.8% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 9.475s  coverage: 2.8% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]

-----------------------------------------------------
--- PASS: TestAccResourceMongoDBAtlasPrivateLinkEndpointServiceADL_importBasic (4.23s)
PASS
coverage: 2.8% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 6.828s  coverage: 2.8% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]

-----------------------------------------------------
--- PASS: TestAccDataSourceMongoDBAtlasPrivateLinkEndpointServiceADL_basic (4.33s)
PASS
coverage: 2.8% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 6.901s  coverage: 2.8% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]

-----------------------------------------------------
--- PASS: TestAccDataSourceMongoDBAtlasPrivateLinkEndpointsServiceADL_basic (4.34s)
PASS
coverage: 3.0% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 6.976s  coverage: 3.0% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]

### CODE REVIEW

**Comment by abner-dou** [comment]
@themantissa There's still a problem with the API_KEYS that actually Edgar is resolving in another PR, that's why this PR is failing in this moment

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	// Get client connection.
+	conn := meta.(*MongoDBClient).Atlas
+	projectID := d.Get("project_id").(string)
+	endpointID := d.Get("endpoint_id").(string)
+
+	privateLinkResponse, _, err := conn.DataLakes.GetPrivateLinkEndpoint(ctx, projectID, endpointID)
+	if err != nil {
+		// case 404
+		// deleted in the backend case
+		reset := strings.Contains(err.Error(), "404") && !d.IsNewResource()
```
Comment (andreaangiolillo) [inline mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go]: [q] Why did you add `&& !d.IsNewResource()`? why getting 404 is not enough to remove the resource?

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoints_service_adl.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+				"endpoint_id":   privateLink.EndpointID,
+				"type":          privateLink.Type,
+				"provider_name": privateLink.Provider,
+				"comment":       privateLink.Comment,
+			}
+		}
+	}
+
+	return results
+}
```
Comment (andreaangiolillo) [inline mongodbatlas/data_source_mongodbatlas_privatelink_endpoints_service_adl.go]:
```suggestion
func flattenADLPrivateLinkEndpoints(privateLinks []*matlas.PrivateLinkEndpointDataLake) []map[string]interface{} {
	var results []map[string]interface{}

	if len(privateLinks) == 0{
		return results
	}

	results = make([]map[string]interface{}, len(privateLinks))

	for k, privateLink := range privateLinks {
		results[k] = map[string]interface{}{
			"endpoint_id":   privateLink.EndpointID,
			"type":          privateLink.Type,
			"provider_name": privateLink.Provider,
			"comment":       privateLink.Comment,
		}
	}
```

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go (hunk 1 above)
Comment (abner-dou) [reply mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go]: this was added to avoid eventual inconsistency but in this case is not necessary

**Review by thetonymaster — APPROVED** [review]

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoints_service_adl.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+func dataSourceMongoDBAtlasPrivateLinkEndpointsServiceADLRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// Get client connection.
+	conn := meta.(*MongoDBClient).Atlas
+	projectID := d.Get("project_id").(string)
+
+	privateLinkEndpoints, _, err := conn.DataLakes.ListPrivateLinkEndpoint(ctx, projectID)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error getting privateLinkEndpoints information: %s", err))
```
Comment (themantissa) [inline mongodbatlas/data_source_mongodbatlas_privatelink_endpoints_service_adl.go]:
```suggestion
		return diag.FromErr(fmt.Errorf("Error getting ADL PrivateLink Endpoints Information: %s", err))
```

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	privateLinkResponse, _, err := conn.DataLakes.GetPrivateLinkEndpoint(ctx, projectID, endpointID)
+	if err != nil {
+		// case 404
+		// deleted in the backend case
+		if strings.Contains(err.Error(), "404") {
+			d.SetId("")
+			return nil
+		}
+
+		return diag.Errorf("error getting adl private link endpoint  information: %s", err)
```
Comment (themantissa) [inline mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go]:
```suggestion
		return diag.Errorf("Error getting ADL PrivateLink Endpoint Information: %s", err)
```

File: mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	matlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const (
+	errorADLServiceEndpointAdd = "error adding MongoDB ADL Private Link Endpoint Connection(%s): %s"
```
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go]:
```suggestion
	errorADLServiceEndpointAdd = "Error adding MongoDB ADL PrivateLink Endpoint Connection(%s): %s"
```

File: mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+func resourceMongoDBAtlasPrivateLinkEndpointServiceADLUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// Get client connection.
+	conn := meta.(*MongoDBClient).Atlas
+	ids := decodeStateID(d.Id())
+	projectID := ids["project_id"]
+	endpointID := ids["endpoint_id"]
+
+	privateLink, _, err := conn.DataLakes.GetPrivateLinkEndpoint(ctx, projectID, endpointID)
+	if err != nil {
+		return diag.Errorf("error getting private link endpoint information: %s", err)
```
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go]: Shouldn't these call out specifically ADL like others?
```suggestion
		return diag.Errorf("Error getting ADL PrivateLink Endpoint Information: %s", err)
```

File: mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	if d.HasChange("comment") {
+		privateLink.Comment = d.Get("comment").(string)
+	}
+
+	_, _, err = conn.DataLakes.CreatePrivateLinkEndpoint(context.Background(), projectID, privateLink)
+	if err != nil {
+		return diag.Errorf("error updating private link endpoint (%s): %s", privateLink.EndpointID, err)
```
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go]:
```suggestion
		return diag.Errorf("Error updating ADL PrivateLink endpoint (%s): %s", privateLink.EndpointID, err)
```

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown [hunk 8]
```diff
@@ -0,0 +1,48 @@
+---
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: privatelink_endpoint_service_adl"
+sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoint-service-adl"
+description: |-
+Describes a DataLake and Online Archive private link endpoint.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
Describes an Atlas Data Lake and Online Archive PrivateLink endpoint
```

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+---
+
+
+# privatelink_endpoint_service_adl
+
+`privatelink_endpoint_service_adl` Provides DataLake and Online Archive private link endpoint resource.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
`privatelink_endpoint_service_adl` Provides an Atlas Data Lake (ADL) and Online Archive PrivateLink endpoint resource.
```

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.
+
+## Example Usage
+
+### Basic
+```terraform
+resource "mongodbatlas_privatelink_endpoint_service_adl" "adl_test" {
+  project_id   = "<PROJECT_ID>"
+  endpoint_id  = "<ENDPOINT_ID>"
+  comment      = "comments for private link endpoint adl"
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
  comment      = "Comment for PrivateLink endpoint ADL"
```

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+## Argument Reference
+
+* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.
+* `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `type` - Human-readable label that identifies the resource associated with this private endpoint. Value is `DATA_LAKE`.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
* `type` - Human-readable label that identifies the type of resource to associate with this private endpoint.
```

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+* `provider_name` - Human-readable label that identifies the cloud provider for this endpoint. Value is AWS.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
* `provider_name` - Human-readable label that identifies the cloud provider for this endpoint. 
```

File: website/docs/d/privatelink_endpoints_service_adl.html.markdown [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+### links
+Each object in the `links` array represents an online archive with the following attributes:
+* `self` - The URL endpoint for this resource.
+
+### results
+
+Each object in the `results` array represents an online archive with the following attributes:
+
+* `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoints_service_adl.html.markdown:53]: Same edits as in singular data source.

File: website/docs/r/privatelink_endpoint_service_adl.html.markdown [hunk 14]
```diff
@@ -0,0 +1,44 @@
+---
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: privatelink_endpoint_service_adl"
+sidebar_current: "docs-mongodbatlas-resource-privatelink-endpoint-service-adl"
+description: |-
+Provides DataLake and Online Archive private link endpoint.
```
Comment (themantissa) [inline website/docs/r/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
Provides an Atlas Data Lake and Online Archive PrivateLink endpoint.
```

File: website/docs/r/privatelink_endpoint_service_adl.html.markdown [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+---
+
+
+# privatelink_endpoint_service_adl
+
+`privatelink_endpoint_service_adl` Describe a DataLake and Online Archive private link endpoint resource.
```
Comment (themantissa) [inline website/docs/r/privatelink_endpoint_service_adl.html.markdown]:
```suggestion
`privatelink_endpoint_service_adl` Provides an Atlas Data Lake (ADL) and Online Archive PrivateLink endpoint resource.   The same configuration will provide a PrivateLink conection for either Atlas Data Lake or Online Archive.  
```

File: website/docs/d/privatelink_endpoints_service_adl.html.markdown [hunk 16]
```diff
@@ -0,0 +1,58 @@
+---
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: privatelink_endpoints_service_adl"
+sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoints-service-adl"
+description: |-
+Describes the list of all DataLake and Online Archive private link endpoints.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoints_service_adl.html.markdown]:
```suggestion
Describes the list of all Atlas Data Lake and Online Archive PrivateLink endpoints.
```

File: website/docs/d/privatelink_endpoints_service_adl.html.markdown [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+---
+
+# privatelink_endpoints_service_adl
+
+`privatelink_endpoints_service_adl` Describes the list of all DataLake and Online Archive private link endpoints.
```
Comment (themantissa) [inline website/docs/d/privatelink_endpoints_service_adl.html.markdown]:
```suggestion
`privatelink_endpoints_service_adl` Describes the list of all Atlas Data Lake (ADL) and Online Archive PrivateLink endpoints.
```

**Review by themantissa — CHANGES_REQUESTED** [review]
Error and docs suggestions/comments.   Once edited I'd like our doc team to review before we merge to ensure that how we've presented this makes sense for both ADL and Online Archive.

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoints_service_adl.go (hunk 3 above)
Comment (thetonymaster) [reply mongodbatlas/data_source_mongodbatlas_privatelink_endpoints_service_adl.go]: Linter complains when using a capital letter at the beggining of an error, second part it's ok though

File: mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go (hunk 4 above)
Comment (abner-dou) [reply mongodbatlas/data_source_mongodbatlas_privatelink_endpoint_service_adl.go]: done!

File: mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go (hunk 5 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go]: done!

File: mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go (hunk 6 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_privatelink_endpoint_service_adl.go]: done!

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown (hunk 8 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown (hunk 9 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown (hunk 10 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown (hunk 11 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoint_service_adl.html.markdown (hunk 12 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoints_service_adl.html.markdown (hunk 13 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoints_service_adl.html.markdown:53]: done!

File: website/docs/r/privatelink_endpoint_service_adl.html.markdown (hunk 14 above)
Comment (abner-dou) [reply website/docs/r/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/r/privatelink_endpoint_service_adl.html.markdown (hunk 15 above)
Comment (abner-dou) [reply website/docs/r/privatelink_endpoint_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoints_service_adl.html.markdown (hunk 16 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoints_service_adl.html.markdown]: done!

File: website/docs/d/privatelink_endpoints_service_adl.html.markdown (hunk 17 above)
Comment (abner-dou) [reply website/docs/d/privatelink_endpoints_service_adl.html.markdown]: done!

**Review by themantissa — APPROVED** [review]
LGTM, will tag docs just for a quick check.

**Comment by themantissa** [comment]
@jwilliams-mongo can you review the docs portion of this for a check?   Per guidence from the Data Lake team we are using the same resource for both Data Lake and Online Archive as it's the exact same API endpoint.  In the Atlas Admin Doc it's described with two different endpoints, which makes it clear but wouldn't really work here.   Does this seem clear enough to you?

**Comment by jwilliams-mongo** [comment]
Sorry for the delay @themantissa I was OOTO. Would you still like me to take a look at this?
